### PR TITLE
Get binding information without cluster connection

### LINKED ISSUE #6420: odo describe binding crash
/kind bug

## What versions of software are you using?

**Operating System:** Win11Pro

**Output of `odo version`:**

`odo v3.4.0 (b8662ef74)`

## How did you run odo exactly?

`odo describe binding`

## Actual behavior

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0x187d19a]

goroutine 1 [running]:
github.com/redhat-developer/odo/pkg/binding.(*BindingClient).GetBindingsFromDevfile(0xc000823e80, {{{0xc00014e528, 0x5}, {0xc00046f3e0, 0x2d}, {0xc00046f380, 0x2d}, {0xc00029ca00, 0x1386, 0x1500}, ...}, ...}, ...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/binding/binding.go:148 +0x3ba
github.com/redhat-developer/odo/pkg/odo/cli/describe.(*BindingOptions).runWithoutName(0xc000937bc0, {0x22924d8, 0xc0008d33e0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:99 +0x105
github.com/redhat-developer/odo/pkg/odo/cli/describe.(*BindingOptions).Run(0x22924d8?, {0x22924d8?, 0xc0008d33e0?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:69 +0x47
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x2292ee8, 0xc000937bc0}, 0xc000933400, {0x34a0428, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/genericclioptions/runnable.go:278 +0x15cd
github.com/redhat-developer/odo/pkg/odo/cli/describe.NewCmdBinding.func1(0xc000933400?, {0x34a0428?, 0x0?, 0x0?})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/pkg/odo/cli/describe/binding.go:117 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000933400, {0x34a0428, 0x0, 0x0})
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:872 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001bcf00)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:918
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/vendor/github.com/spf13/cobra/command.go:911
main.main()
        /builddir/build/BUILD/gocode/src/github.com/redhat-developer/odo/cmd/odo/odo.go:83 +0x579
```

## Expected behavior

Binding should be listed

## Any logs, error output, etc?

[devfile.zip](https://github.com/redhat-developer/odo/files/10246319/devfile.zip)

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind bug

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6420 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

**Comment by feloy** [comment]
When running `odo describe component` without cluster access, we display information of the bindings defined in the Devfile.

odo is using `corev1.ObjectReference` structure to give information about the `Applciation` and `Service` defining a binding.

With `ObjectReference`, the resource is defined with `APIVersion` and `Kind`.

```
type ObjectReference struct {
	Kind string
	Namespace string
	Name string
	UID types.UID
	APIVersion string
	ResourceVersion string
	FieldPath string
}
```

In the other hand, in the Devfile, a binding can be defined either by a `ServiceBinding.binding.operators.coreos.com/v1alpha1` or a `ServiceBinding.servicebinding.io/v1alpha3`.

When using the first binding CRD (`ServiceBinding.binding.operators.coreos.com/v1alpha1`), the resource of the `Applciation` can be defined either by its `Kind` or `Resource`.

When an Application is defined with a Resource (as opposed to kind), we need access to the cluster to use the RESTMapper to get the Kind from the Resource.

But without cluster access, access to RESTMapper fails.

File: tests/integration/cmd_describe_list_binding_test.go [hunk 1]
```diff
@@ -41,48 +41,52 @@ var _ = Describe("odo describe/list binding command tests", func() {
...
-						Expect(stdout).To(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io)"))
-						Expect(stdout).ToNot(ContainSubstring("cluster-sample (Cluster.postgresql.k8s.enterprisedb.io) (namespace: "))
-					}
-					Expect(stdout).To(ContainSubstring("Bind as files: true"))
-					Expect(stdout).To(ContainSubstring("Detect binding resources: true"))
-					Expect(stdout).To(ContainSubstring("Naming strategy: lowercase"))
-					Expect(stdout).To(ContainSubstring("Available binding information: unknown"))
-					Expect(stdout).To(ContainSubstring("Binding information for one or more ServiceBinding is not available"))
+			for _, label := range []string{
+				helper.LabelNoCluster, helper.LabelUnauth,
```
Comment (rm3l) [inline tests/integration/cmd_describe_list_binding_test.go:45]: So this no longer tests the case where there is a cluster, no? I was thinking of introducing an explicit "Cluster" label for such case. What do you think?

**Review by rm3l — CHANGES_REQUESTED** [review]
LGTM overall - thanks for fixing this. I added a small comment regarding the tests.
And I had a few more:

1) When running `odo describe binding` with no cluster, I was expecting a  warning (like `unable to access the cluster` or `No connection to cluster defined`) to be displayed, but it was not. Can we display such a message, to be consistent with other commands like `odo describe component`?

2) I noticed that we had a similar panic with `odo list binding` when there is no cluster. Can we fix it also here (since this PR is about "getting binding info without cluster connection"). Or, if you prefer, it can be addressed in a separate issue/PR.
```
$ KUBECONFIG=/dev/null odo list binding      
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xf0 pc=0x299d1d2]

goroutine 1 [running]:
github.com/redhat-developer/odo/pkg/odo/cli/list/binding.(*BindingListOptions).Run(0xc00081c840, {0x3460ac8, 0xc000ab72c0})
        /home/asoro/work/projects/odo/pkg/odo/cli/list/binding/binding.go:71 +0x72
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x3461580, 0xc00081c840}, 0xc000bef180, {0x479f308, 0x0, 0x0})
        /home/asoro/work/projects/odo/pkg/odo/genericclioptions/runnable.go:280 +0x1619
github.com/redhat-developer/odo/pkg/odo/cli/list/binding.NewCmdBindingList.func1(0xc000bef180?, {0x479f308?, 0x0?, 0x0?})
        /home/asoro/work/projects/odo/pkg/odo/cli/list/binding/binding.go:115 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000bef180, {0x479f308, 0x0, 0x0})
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:872 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc0005dca00)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:918
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:911
main.main()
        /home/asoro/work/projects/odo/cmd/odo/odo.go:83 +0x579
```

File: tests/integration/cmd_describe_list_binding_test.go (hunk 1 above)
Comment (feloy) [reply tests/integration/cmd_describe_list_binding_test.go:45]: These tests are when odo dev is not running, I expect it is enough to run them without cluster

**Comment by feloy** [comment]
@rm3l I worked on 2. 

For 1., I think it would need some specific work to clearly specify what would be the outputs for the different situations. With podman, it adds some complexity on what we would like to have as warnings.

File: tests/integration/cmd_describe_list_binding_test.go (hunk 1 above)
Comment (rm3l) [reply tests/integration/cmd_describe_list_binding_test.go:45]: Indeed, that makes sense.

**Review by rm3l — COMMENTED** [review]
Thanks for fixing the issue with "odo list binding" as well!
I think there is still one issue when using the "--namespace" flag, which is causing a panic because of the nilable `KubernetesClient`:

```
$ KUBECONFIG=/dev/null odo list binding --namespace 404
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2a8 pc=0x299d0b6]

goroutine 1 [running]:
github.com/redhat-developer/odo/pkg/odo/cli/list/binding.(*BindingListOptions).Complete(0x3460a68?, {0xc000a9e960?, 0x2d0fc00?}, {0x479f308?, 0x2be3d60?}, {0x47a1b00?, 0x0?, 0x0?})
        /home/asoro/work/projects/odo/pkg/odo/cli/list/binding/binding.go:59 +0x36
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x3461520, 0xc000b97218}, 0xc000ca7180, {0xc000cbcec0, 0x0, 0x2})
        /home/asoro/work/projects/odo/pkg/odo/genericclioptions/runnable.go:261 +0x14a6
github.com/redhat-developer/odo/pkg/odo/cli/list/binding.NewCmdBindingList.func1(0xc000ca7180?, {0xc000cbcec0?, 0x2?, 0x2?})
        /home/asoro/work/projects/odo/pkg/odo/cli/list/binding/binding.go:115 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000ca7180, {0xc000cbcea0, 0x2, 0x2})
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:872 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc000c94a00)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:990 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:918
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:911
main.main()
        /home/asoro/work/projects/odo/cmd/odo/odo.go:83 +0x579
```

**Comment by rm3l** [comment]
> @rm3l I worked on 2.
> 
> For 1., I think it would need some specific work to clearly specify what would be the outputs for the different situations. With podman, it adds some complexity on what we would like to have as warnings.

Thanks. Yes, I agree we need to clarify this. Okay, we can discuss this topic later, maybe in a Cabal call.

**Review by rm3l — APPROVED** [review]
Thanks!

**Review by rm3l — APPROVED** [review]

**Comment by rm3l** [comment]
/override ci/prow/v4.11-integration-e2e

Similar issue reported in https://github.com/redhat-developer/odo/issues/6463.
And all tests are passing on IBM Cloud.

**Pull request merged**